### PR TITLE
multi-modality model construction support

### PR DESCRIPTION
**Pull request by Gasoonjia**
This PR makes torchchat support multi-modality model definition and constructions. To show our power in multi-modality area, we integrate flamingo component into our system.
Note that this is only for bare-minimum support for model definition. Please check [openai_api_multimodal](https://github.com/pytorch/torchchat/tree/openai_api_multimodal) branch for e2e, and https://github.com/pytorch/torchchat/pull/1123#issuecomment-2339028401 for better structure and llama3.1 support

### CODE REVIEW

File: torchchat/model.py [hunk 1]
```diff
@@ -77,13 +124,32 @@ def from_params(cls, params):
                 params[_to] = params.pop(_from)
         return cls(**params)
 
+
 @dataclass
 class ModelArgs:
```
Comment (Jack-Khuu) [inline torchchat/model.py:129]: Docstring

File: torchchat/model.py [hunk 2]
```diff
@@ -184,13 +250,48 @@ class Model(nn.Module):
...
-        return self.text_transformer(idx, input_pos)
+        # TODO: unify the model init logic
+        if config.model_type == ModelType.TextOnly:
+            self.text_transformer = Transformer(config.transformer_args["text"])
+        else:
+            self.model = self.build_model()
     
-    def setup_caches(self, max_batch_size, max_seq_length):
-        self.text_transformer.setup_caches(max_batch_size, max_seq_length)
+    def build_model(self):
```
Comment (Jack-Khuu) [inline torchchat/model.py:259]: This is where all the magic comes together; Let's add a docstring here

File: torchchat/model.py [hunk 3]
```diff
@@ -184,13 +250,48 @@ class Model(nn.Module):
     def __init__(self, config: ModelArgs) -> None:
```
Comment (Jack-Khuu) [inline torchchat/model.py:250]: Since we have the legacy `text_transformer` and new `model`, let's add a quick description until we unify them later

Comment (Jack-Khuu) [reply torchchat/model.py:250]: Or link to your other PR where you fix this

**Review by Jack-Khuu — APPROVED** [review]
This is looking great!! Thanks for making it so easy to review

Mostly nits, but I know this is part of a PR stack we're landing, so we have some leeway

**Comment by Jack-Khuu** [comment]
Before landing:

- [ ] update the PR title and description
- [ ] Check that the existing text only model flow works (walk through README)

File: .DS_Store
[the remark this answers is not in the document]
Comment (Jack-Khuu) [reply .DS_Store:1]: There's a few DS Stores

File: torchchat/model.py [hunk 4]
```diff
@@ -77,13 +124,32 @@ def from_params(cls, params):
...
+    def __init__(
+        self,
+        transformer_args: Union[TransformerArgs, Dict[str, TransformerArgs]],
+        model_type: ModelType = ModelType.TextOnly,
+    ) -> None:
+        self._sanity_check(transformer_args, model_type)
+        
+        self.model_type = model_type
+        if isinstance(transformer_args, TransformerArgs):
+            self.transformer_args = {"text": transformer_args}
```
[the remark this answers is not in the document]
Comment (Gasoonjia) [reply torchchat/model.py:142]: maybe in the different PR if you think it is good. I have some plans to make the configuration more concise and structual.

File: torchchat/model.py (hunk 3 above)
Comment (Gasoonjia) [reply torchchat/model.py:250]: That will happen adjacent to this PR (3.1 torchtune one), so i'd like to keep as it is here.

File: torchchat/model.py [hunk 5]
```diff
@@ -77,13 +124,32 @@ def from_params(cls, params):
                 params[_to] = params.pop(_from)
         return cls(**params)
 
+
 @dataclass
 class ModelArgs:
-    text_transformer_args: TransformerArgs
+    model_type: ModelType
+    transformer_args: Dict[str, Union[Dict, TransformerArgs]]
```
[the remark this answers is not in the document]
Comment (Gasoonjia) [reply torchchat/model.py:131]: No, it is up-to-date.
For torchchat native builder, we need to have TransformerArgs. 
For torchtune, we need Dict.

File: dist_run.py [hunk 6]
```diff
@@ -122,7 +122,7 @@ def main():
     gpu_memory_monitor = GPUMemoryMonitor("cuda")
     logger.info(f"{color.yellow} {gpu_memory_monitor.get_device_info()}{color.reset}")
 
-    config = ModelArgs.from_name(MODEL_NAME).text_transformer_args
+    config = ModelArgs.from_name(MODEL_NAME)..transformer_args['text']
```
Comment (Jack-Khuu) [inline dist_run.py:125]: Double dot

File: torchchat/model.py [hunk 7]
```diff
@@ -42,22 +42,41 @@ class ModelType(Enum):
     TextOnly = "text_only"
     Flamingo = "flamingo"
 
+# Type for objects that can generate nn.Module instance
+ModuleLike = Union[nn.Module, Callable[..., nn.Module]]
 
 @dataclass
 class ModelRecipe:
+    """
+    A class in TorchChat that describes and contains all supported model structures in TorchChat.
```
Comment (Jack-Khuu) [inline torchchat/model.py:51]:
```suggestion
    A class in torchchat that describes and contains all supported model structures in torchchat.
```

File: torchchat/model.py [hunk 8]
```diff
@@ -42,22 +42,41 @@ class ModelType(Enum):
...
+    
+    ModelRecipe represents a model as a collection of Transformer modules and a fusion module,
+    providing a standardized and centralized way to define and build models in TorchChat.
```
Comment (Jack-Khuu) [inline torchchat/model.py:54]:
```suggestion
    providing a standardized and centralized way to define and build models in torchchat.
```

File: torchchat/model.py [hunk 9]
```diff
@@ -247,6 +267,9 @@ def update(self, input_pos, k_val, v_val):
 
 
 class Model(nn.Module):
+    """
+    The entrance for model construction in tochchat.
```
Comment (Jack-Khuu) [inline torchchat/model.py:271]:
```suggestion
    The entrance for model construction in torchchat.
```

**Review by Jack-Khuu — APPROVED** [review]

**Pull request merged**